src/Balsam/src/Balsam.Api/Controllers/WorkspaceController.cs

```diff
@@ -97,5 +94,61 @@ public override Task<IActionResult> ListTemplates()
                 Title = "No workspace template found"
             }));
         }
+
+        public async override Task<IActionResult> ListWorkspaces([FromQuery(Name = "projectId")] string? projectId, [FromQuery(Name = "branchId")] string? branchId, [FromQuery(Name = "all")] bool? all)
+        {
+            if (branchId != null && projectId is null)
+            {
+                return BadRequest(new Problem() { Status = 400, Type = "Parameter error", Title = "If BranchId is specified a ProjectId must also be specified"});
+            }
+            List<BalsamWorkspace> workspaces;
+            if (projectId != null)
+            {
+                if (all??true)
+                {
+                    if (branchId is null)
+                    {
+                        workspaces = await _hubClient.GetWorkspacesByProject(projectId);
+                    }
+                    else
+                    {
+                        workspaces = await _hubClient.GetWorkspacesByProjectAndBranch(projectId, branchId);
+                    }
+                }
+                else
+                {
+                    var project = await _hubClient.GetProject(projectId);
+
+                    if (this.User.Claims.FirstOrDefault(c => c.Type == "groups" && c.Value == project.Oidc.GroupId) is null)
+                    {
+                        return BadRequest(new Problem() { Status = 400, Type = "Not authorized", Title = "You are not a member of the project" });
+                    }
+
+                    if (branchId is null)
+                    {
+                        workspaces = await _hubClient.GetWorkspacesByProjectAndUser(projectId, this.User.Claims.FirstOrDefault(c => c.Type == "preferred_username")?.Value);

```

Comment:
börjar bli många träffar i filen på:
`this.User.Claims.FirstOrDefault(c => c.Type == "preferred_username")?.Value`

kanske skapa private method för getUserName typ:
```
private string getUserName()
{
    return User?.Claims?.FirstOrDefault(c => c.Type == "preferred_username")?.Value;
}
```